bigtable/client/detail/bulk_mutator.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BIGTABLE_CLIENT_DETAIL_BULK_MUTATOR_H_
+#define BIGTABLE_CLIENT_DETAIL_BULK_MUTATOR_H_
+
+#include <bigtable/client/idempotent_mutation_policy.h>
+
+#include <google/bigtable/v2/bigtable.grpc.pb.h>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+namespace detail {
+/// Keep the state in the Table::BulkApply() member function.
+class BulkMutator {
+ public:
+  BulkMutator(std::string const& table_name,
+              IdempotentMutationPolicy& idempotent_policy, BulkMutation&& mut);
+
+  /// Return true if there are pending mutations in the mutator
+  bool HasPendingMutations() const {
+    return pending_mutations.entries_size() != 0;
+  }
+
+  /// Send one batch request to the given stub.
+  grpc::Status MakeOneRequest(
+      google::bigtable::v2::Bigtable::StubInterface& stub,
+      grpc::ClientContext& client_context);
+
+  /// Give up on any pending mutations, move them to the failures array.
+  std::vector<FailedMutation> ExtractFinalFailures();
+
+ private:
+  /// Get ready for a new request
+  void PrepareForRequest();
+
+  /// Process a single response.
+  void ProcessResponse(google::bigtable::v2::MutateRowsResponse& response);
+
+  /// A request has finished and we have processed all the responses.
+  void FinishRequest();
+
+ private:
```

Comment:
Should private member vars have a trailing underscore?

Reply:
Yup.  Fixed.